### PR TITLE
feat(toolgen): align MCP tool generation with best practices

### PR DESCRIPTION
**Pull request by devantler**
Aligns KSail's MCP tool generation with patterns used by production MCP servers (github-mcp-server, modelcontextprotocol/servers filesystem/memory). The main motivation is reducing the token footprint of tool schemas -- KSail's 6 tools previously consumed ~86KB (~21,650 tokens), more than GitHub's MCP server uses for 100 tools (~63KB) -- and enabling MCP client features like auto-approving reads and warning on destructive writes.

## Approach

Six focused changes, ordered by impact:

1. **MCP annotations** -- Add `Title`, `ReadOnlyHint`, `DestructiveHint`, `IdempotentHint`, and `OpenWorldHint` to every generated tool. Read tools get `readOnly=true, idempotent=true`; write tools get `destructive=true`. All tools set `openWorld=true` (KSail can target remote clusters, cloud providers such as Hetzner and AWS EKS, and OCI registries). This is the minimum bar set by github-mcp-server and the filesystem reference server.

2. **Description truncation** -- Cap verbose kubectl flag descriptions at 200 chars with sentence-boundary-aware truncation. Some descriptions were over 1,000 chars (e.g., the `validate` flag at 1,109 chars). A total cap of 400 chars is enforced after appending "(applies to: ...)" suffixes.

3. **Applies-to condensing** -- Use a 50% threshold for the `(applies to: ...)` suffix on consolidated tool parameters. Flags that apply to most subcommands no longer list them all. This alone saves thousands of chars in `workload_write`.

4. **Flag exclusion** -- Hide 16 rarely-used kubectl config flags (`cache-dir`, `certificate-authority`, `client-certificate`, `client-key`, `server`, `tls-server-name`, `insecure-skip-tls-verify`, `insecure-skip-tls-verify-backend`, `disable-compression`, `as`, `as-group`, `as-uid`, `as-user-extra`, `request-timeout`, `token`, `user`) from the AI schema. The flags still work if explicitly passed.

5. **Enum compaction** -- Cap subcommand enum descriptions at 60 chars in the consolidated tool's command selector.

6. **Server instructions** -- Replace the generic one-liner with workflow-oriented text describing the init-create-update-delete lifecycle and tool responsibilities.

## Results

| Metric | Before | After |
|--------|--------|-------|
| Total schema bytes | 86,601 | 71,032 |
| Estimated tokens | ~21,650 | ~17,758 |
| Token reduction | -- | **~18%** (~3,892 tokens) |
| MCP annotation coverage | 0% | 100% |
| Client auto-approve reads | No | Yes |
| Client destructive warnings | No | Yes |

## Type of change

Select the appropriate options and delete the rest:

- [x] 🧹 Refactor

### CODE REVIEW

File: pkg/toolgen/schema.go [hunk 1]
```diff
@@ -339,9 +351,12 @@ func addFlagProperties(
...
-		// Build description with conditional applicability
-		description := flagDef.Description
-		if len(flagDef.AppliesToSubcommands) < totalSubcommands {
+		// Build description with conditional applicability.
+		// Truncate the base description first, then append "(applies to: ...)" if needed.
+		description := truncateDescription(flagDef.Description, maxDescriptionLength)
+
+		applicability := float64(len(flagDef.AppliesToSubcommands)) / float64(totalSubcommands)
+		if applicability < appliesToThreshold {
 			description = fmt.Sprintf(
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go:397]: Fixed. Added `maxTotalFlagDescriptionLength = 400` constant and a `truncateDescription` call after appending the "(applies to: ...)" suffix to enforce an upper bound on the final description length.

File: pkg/toolgen/options.go [hunk 2]
```diff
@@ -100,6 +104,25 @@ func DefaultOptions() ToolOptions {
 			"ksail help",       // Help command, not a tool
 			"ksail",            // Root command, not a tool
 		},
+		ExcludeFlags: []string{
+			// kubectl global config flags rarely used by LLMs.
+			// These are still accepted at runtime — just hidden from the AI schema.
+			"cache-dir",
+			"certificate-authority",
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/options.go:111]: Fixed. Added `"request-timeout"` to the ExcludeFlags list (now 16 flags total).

File: pkg/toolgen/definition.go [hunk 3]
```diff
@@ -34,6 +42,26 @@ type ToolDefinition struct {
...
+type ToolAnnotationHints struct {
+	// ReadOnlyHint indicates the tool does not modify its environment.
+	ReadOnlyHint bool
+
+	// DestructiveHint indicates the tool may perform destructive updates.
+	// Only meaningful when ReadOnlyHint is false.
+	DestructiveHint bool
+
+	// IdempotentHint indicates repeated calls with the same args have no additional effect.
+	// Only meaningful when ReadOnlyHint is false.
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/definition.go]: Fixed. Updated the IdempotentHint comment to: "Set true for both read-only tools (reads are inherently idempotent) and write tools whose operations are idempotent (e.g., declarative apply)."

File: pkg/toolgen/definition.go [hunk 4]
```diff
@@ -34,6 +42,26 @@ type ToolDefinition struct {
 	Subcommands map[string]*SubcommandDef
 }
 
+// ToolAnnotationHints contains SDK-agnostic behavioral hints for a tool.
+// These map to MCP ToolAnnotations and inform client behavior.
+// All fields follow the MCP spec defaults: unset means worst-case assumptions.
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/definition.go]: Fixed. Updated the comment to say "KSail always sets all fields explicitly so there is no ambiguity." -- plain bool is intentional since we always set explicit values and don't need tri-state semantics.

File: pkg/toolgen/schema.go [hunk 5]
```diff
@@ -429,3 +450,30 @@ func mapFlagTypeToItemsType(flagType string) string {
...
+		}
+	}
+
+	// Try to break at a space
+	if idx := strings.LastIndex(desc[:cutoff], " "); idx > 0 {
+		return desc[:idx] + "…"
+	}
+
+	// Hard cut
+	return desc[:cutoff] + "…"
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go]: Fixed. Switched from Unicode `\u2026` (3 bytes) to ASCII `"..."` (3 bytes, 3 chars). Reservation is now `maxLen - len(ellipsis)` where `ellipsis = "..."`. Also added a guard for very small `maxLen` values (returns just `"..."` if maxLen <= 3). All covered by new unit tests with a `LessOrEqual` assertion to verify the invariant.

File: pkg/toolgen/schema.go [hunk 6]
```diff
@@ -429,3 +444,30 @@ func mapFlagTypeToItemsType(flagType string) string {
 		return ""
 	}
 }
+
+// truncateDescription truncates a description string to maxLen characters.
+// It tries to break at a sentence boundary ('. ') or clause boundary (', ', '; ')
+// within the limit. Falls back to a hard cut with '…' suffix.
+func truncateDescription(desc string, maxLen int) string {
+	if len(desc) <= maxLen {
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go]: Fixed. Added `TestTruncateDescription` in `schema_test.go` with 8 test cases covering: no truncation, exact length, sentence boundary, clause boundary, space break, hard cut (no spaces), very small maxLen, and a `LessOrEqual` assertion to verify the `maxLen` invariant holds for all cases. Exported via `export_test.go` seam.

File: pkg/toolgen/options.go [hunk 7]
```diff
@@ -100,6 +104,25 @@ func DefaultOptions() ToolOptions {
...
+			"insecure-skip-tls-verify-backend",
+			"tls-server-name",
+			"as",
+			"as-group",
+			"as-uid",
+			"as-user-extra",
+			"token",
+			"server",
+			"user",
+		},
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/options.go:126]: Fixed. Added `"request-timeout"` to ExcludeFlags. The `insecure-skip-tls-verify-backend` flag is intentionally excluded -- it's a niche kubectl flag for skipping kubelet TLS verification during log fetching, almost never needed by LLMs. Will update PR description to reflect the actual list.

File: pkg/toolgen/mcp.go [hunk 8]
```diff
@@ -73,6 +78,25 @@ func addMCPTool(server *mcp.Server, tool ToolDefinition, opts ToolOptions) {
...
+
+// buildMCPAnnotations creates MCP ToolAnnotations from the tool's behavioral hints.
+func buildMCPAnnotations(tool ToolDefinition) *mcp.ToolAnnotations {
+	return &mcp.ToolAnnotations{
+		Title:           tool.Title,
+		ReadOnlyHint:    tool.Annotations.ReadOnlyHint,
+		DestructiveHint: new(tool.Annotations.DestructiveHint),
+		IdempotentHint:  tool.Annotations.IdempotentHint,
+		OpenWorldHint:   new(tool.Annotations.OpenWorldHint),
+	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/mcp.go]: This compiles and works correctly in Go 1.26. The `new(expr)` syntax was introduced alongside `//go:fix inline` support -- `new` can now accept a value expression and returns a pointer to a copy of that value. The CI build and all tests pass with this syntax. Removed the dead `boolPtr` wrapper; the inlined calls remain.

File: pkg/toolgen/mcp.go [hunk 9]
```diff
@@ -73,6 +78,23 @@ func addMCPTool(server *mcp.Server, tool ToolDefinition, opts ToolOptions) {
 	mcp.AddTool(server, mcpTool, handler)
 }
 
+// boolPtr returns a pointer to a bool value.
+// MCP SDK uses *bool to distinguish "explicitly false" from "unset" (nil).
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/mcp.go]: Fixed. Removed `boolPtr` entirely. The `buildMCPAnnotations` comment now explains why `DestructiveHint`/`OpenWorldHint` use `*bool` (nil = spec default) while `ReadOnlyHint`/`IdempotentHint` are plain bool (zero value = same as spec default).

File: pkg/toolgen/mcp.go [hunk 10]
```diff
@@ -73,6 +78,25 @@ func addMCPTool(server *mcp.Server, tool ToolDefinition, opts ToolOptions) {
...
+//
+//go:fix inline
+func boolPtr(b bool) *bool {
+	return new(b)
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/mcp.go]: Fixed. Removed both `boolPtr` and the `//go:fix inline` directive entirely. The CI auto-fixer had inlined the calls using Go 1.26's `new(expr)` syntax, leaving the function as dead code. The inlined `new(expr)` calls in `buildMCPAnnotations` are valid Go 1.26 and remain.

File: pkg/toolgen/schema.go [hunk 11]
```diff
@@ -284,7 +299,10 @@ func buildSubcommandEnumProperty(subcommands map[string]*SubcommandDef) map[stri
 
 	for name, def := range subcommands {
 		names = append(names, name)
-		descriptions = append(descriptions, name+": "+def.Description)
+		descriptions = append(
+			descriptions,
+			name+": "+truncateDescription(def.Description, maxSubcommandDescLength),
+		)
 	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go]: Fixed. Refactored `buildSubcommandEnumProperty` to collect keys first, sort them with `slices.Sort`, then build both the `enum` slice and `descriptions` slice in that sorted order. Also sorted `AppliesToSubcommands` in `mergeSubcommandFlags` (after the merge loop) for consistent flag description ordering.

File: pkg/svc/mcp/server.go [hunk 12]
```diff
@@ -50,10 +50,15 @@ func NewServer(cfg ServerConfig) (*mcpsdk.Server, error) {
...
-		Instructions: "KSail MCP server - provides access to KSail Kubernetes cluster management commands",
-		Logger:       cfg.Logger,
-		KeepAlive:    serverKeepAlive,
-		PageSize:     serverPageSize,
+		Instructions: "KSail — SDK for managing local Kubernetes clusters. Only Docker is required.\n\n" +
+			"Workflow: init → create → [update|info|diagnose] → delete.\n" +
+			"Tools: cluster_read (inspect), cluster_write (lifecycle), " +
+			"workload_read (get/describe/logs), workload_write (apply/create/scale/rollout), " +
+			"cipher_write (SOPS encryption), tenant_write (multi-tenancy).\n" +
+			"All operations target the current kubeconfig context. Use the 'context' parameter to target a different cluster.",
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/svc/mcp/server.go]: Fixed. Replaced the inaccurate blanket statement with tool-specific guidance: "Cluster lifecycle tools use `--name` and optionally `--kubeconfig` to target a cluster. Workload tools accept `--context` to target a kubeconfig context and `--namespace` for namespace scoping."

File: pkg/toolgen/generator.go [hunk 13]
```diff
@@ -172,10 +173,39 @@ func commandToToolDefinition(cmd *cobra.Command) ToolDefinition {
...
+			OpenWorldHint:   false,
+		}
+	}
+
+	return ToolAnnotationHints{
+		ReadOnlyHint:    false,
+		DestructiveHint: true,
+		IdempotentHint:  false,
+		OpenWorldHint:   false,
 	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/generator.go:211]: Fixed in commit `2c3636513`. Changed `OpenWorldHint` from `false` to `true` for both read and write tools. KSail can target external cloud providers (Hetzner, AWS EKS, Omni), remote kubeconfig contexts, and OCI registries — so closed-world (`false`) was misleading. `true` is also the MCP spec default, so this prevents clients from silently auto-approving external-world operations. Also updated the `OpenWorldHint` comment in `definition.go` to document the rationale. `IdempotentHint` for write tools stays `false` as the safe default (not all write ops are idempotent, e.g. `cluster create`/`delete`).

File: pkg/toolgen/schema.go [hunk 14]
```diff
@@ -280,11 +295,18 @@ func buildConsolidatedParameterSchema(
...
+
+	descriptions := make([]string, 0, len(subcommands))
+	for _, name := range names {
+		descriptions = append(
+			descriptions,
+			name+": "+truncateDescription(subcommands[name].Description, maxSubcommandDescLength),
+		)
 	}
 
 	return map[string]any{
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go:313]: Fixed in a prior commit (`a56ebc480`). `buildSubcommandEnumProperty` now collects all map keys into a `names` slice, sorts it with `slices.Sort`, and then builds both the `enum` and `descriptions` slices in that sorted order — so schema output is fully deterministic.

File: pkg/svc/mcp/server.go [hunk 15]
```diff
@@ -50,10 +50,16 @@ func NewServer(cfg ServerConfig) (*mcpsdk.Server, error) {
...
-		Logger:       cfg.Logger,
-		KeepAlive:    serverKeepAlive,
-		PageSize:     serverPageSize,
+		Instructions: "KSail — SDK for managing local Kubernetes clusters. Only Docker is required.\n\n" +
+			"Workflow: init → create → [update|info|diagnose] → delete.\n" +
+			"Tools: cluster_read (inspect), cluster_write (lifecycle), " +
+			"workload_read (get/describe/logs), workload_write (apply/create/scale/rollout), " +
+			"cipher_write (SOPS encryption), tenant_write (multi-tenancy).\n" +
+			"Cluster lifecycle tools use --name and optionally --kubeconfig to target a cluster. " +
+			"Workload tools accept --context to target a kubeconfig context and --namespace for namespace scoping.",
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/svc/mcp/server.go:59]: Fixed in a prior commit (`a56ebc480`). The instructions now explicitly distinguish cluster targeting (`--name`/`--kubeconfig`) from workload targeting (`--context`/`--namespace`): "Cluster lifecycle tools use --name and optionally --kubeconfig to target a cluster. Workload tools accept --context to target a kubeconfig context and --namespace for namespace scoping."

File: pkg/toolgen/constants.go [hunk 16]
```diff
@@ -42,6 +42,27 @@ const (
...
 
+// Token optimization limits.
+const (
+	// maxDescriptionLength is the maximum length for parameter descriptions.
+	// Verbose kubectl help text is truncated to this limit.
+	maxDescriptionLength = 200
+
+	// maxSubcommandDescLength is the maximum length for subcommand descriptions
+	// in the consolidated enum property.
+	maxSubcommandDescLength = 60
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/constants.go:55]: Fixed in commit `2c3636513`. Updated the `maxDescriptionLength` comment to read: "base maximum length for parameter descriptions before any suffix is appended. Verbose kubectl help text is truncated to this limit; the final description may be longer (up to `maxTotalFlagDescriptionLength`) when an '(applies to: ...)' suffix is added for consolidated tools." This makes the base vs. total distinction explicit.

File: pkg/toolgen/schema.go [hunk 17]
```diff
@@ -320,6 +342,11 @@ func mergeSubcommandFlags(subcommands map[string]*SubcommandDef) map[string]*Fla
 		}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go:364]: Fixed in a prior commit (`a56ebc480`). `mergeSubcommandFlags` now calls `slices.Sort(flagDef.AppliesToSubcommands)` after the merge loop, ensuring the "applies to" list is always in lexicographic order regardless of map iteration order.

File: pkg/toolgen/generator.go [hunk 18]
```diff
@@ -172,10 +173,39 @@ func commandToToolDefinition(cmd *cobra.Command) ToolDefinition {
...
+
+// buildToolTitle generates a human-readable title from a tool path.
+// Example: "cluster create" -> "Cluster Create", "workload_read" -> "Workload Read".
+func buildToolTitle(strippedPath string) string {
+	return strutil.SnakeCaseToTitle(strippedPath)
+}
+
+// buildAnnotationHints creates behavioral hints based on permission level.
+// Read-only tools are marked safe; write tools are marked destructive and non-idempotent.
+// All KSail tools operate on local Docker/k8s, so openWorldHint is always false.
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/generator.go]: Fixed in commit `f6f2a86ee`. Updated the comment from "openWorldHint is always false" to "All KSail tools can interact with external entities (remote clusters, cloud providers, OCI registries), so openWorldHint is always true." — matching the actual implementation.

File: pkg/toolgen/generator.go [hunk 19]
```diff
@@ -172,10 +173,39 @@ func commandToToolDefinition(cmd *cobra.Command) ToolDefinition {
...
+			OpenWorldHint:   true,
+		}
+	}
+
+	return ToolAnnotationHints{
+		ReadOnlyHint:    false,
+		DestructiveHint: true,
+		IdempotentHint:  false,
+		OpenWorldHint:   true,
 	}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/generator.go:211]: Fixed in commit `f6f2a86ee`. Updated the PR description to say `openWorld=true` with the rationale: "KSail can target remote clusters, cloud providers such as Hetzner and AWS EKS, and OCI registries." This matches both `buildAnnotationHints` and the comment in `definition.go`.

File: pkg/toolgen/schema.go [hunk 20]
```diff
@@ -429,3 +461,38 @@ func mapFlagTypeToItemsType(flagType string) string {
...
+// It tries to break at a sentence boundary ('. ') or clause boundary (', ', '; ')
+// within the limit. Falls back to a hard cut with "..." suffix.
+func truncateDescription(desc string, maxLen int) string {
+	if len(desc) <= maxLen {
+		return desc
+	}
+
+	// Guard against very small maxLen values
+	if maxLen <= len(ellipsis) {
+		return ellipsis
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go]: Fixed in commit `f6f2a86ee`. Changed the guard from `return ellipsis` to `return ellipsis[:maxLen]` so the function never returns a string longer than `maxLen` bytes. Added test cases for `maxLen=0`, `1`, and `2` to cover these edge cases explicitly.

File: pkg/toolgen/schema.go [hunk 21]
```diff
@@ -225,14 +235,19 @@ func buildEnumProperty(ev enumValuer, flag *pflag.Flag) map[string]any {
...
 
 	cmd.Flags().VisitAll(func(flag *pflag.Flag) {
 		if flag.Name == helpFlagName {
 			return
 		}
 
+		// Skip excluded flags
+		if slices.Contains(excludeFlags, flag.Name) {
+			return
+		}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go]: Fixed in commit `42866f1a1`. `extractFlags` no longer filters excluded flags — `SubcommandDef.Flags` now contains ALL flags. The `excludeFlags` filter was moved to `mergeSubcommandFlags` (schema layer only), so excluded flags remain in `SubcommandDef.Flags` and `handleConsolidatedTool` forwards them at runtime. Updated the test to assert `server` IS in `encryptSubcmd.Flags` (runtime forwarding works) and NOT in the schema properties.

File: pkg/toolgen/definition.go [hunk 22]
```diff
@@ -6,6 +6,10 @@ type ToolDefinition struct {
 	// Name is the tool identifier (e.g., "ksail_cluster_create").
 	Name string
 
+	// Title is a human-readable display name for the tool (e.g., "Cluster Management").
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/definition.go]: Fixed in commit `42866f1a1`. Updated the `ToolDefinition.Title` docstring example from "Cluster Management" to "Cluster Create" / "Workload Read" to match the actual generated titles.

File: pkg/toolgen/options.go [hunk 23]
```diff
@@ -27,6 +27,10 @@ const (
 type ToolOptions struct {
 	// ExcludeCommands is a list of command paths to exclude (e.g., "ksail chat").
 	ExcludeCommands []string
+	// ExcludeFlags is a list of flag names to exclude from tool schemas.
+	// These flags still work at runtime — they're just hidden from the AI schema
+	// to reduce token usage.
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/options.go:32]: Fixed in commit `42866f1a1`. The comment was previously inaccurate — excluded flags were NOT forwarded at runtime for consolidated tools. The root cause was that `extractFlags` removed them from `SubcommandDef.Flags`, making `handleConsolidatedTool` drop them. The fix moves filtering to `mergeSubcommandFlags` (schema only), so excluded flags now genuinely work at runtime as the comment claimed.

File: pkg/toolgen/schema.go [hunk 24]
```diff
@@ -295,15 +318,26 @@ func buildSubcommandEnumProperty(subcommands map[string]*SubcommandDef) map[stri
...
 
 // mergeSubcommandFlags collects all flags from subcommands, tracking which subcommands each applies to.
+// excludeFlags lists flags to omit from the merged result (schema-level exclusion); these flags
+// remain in each SubcommandDef.Flags so handleConsolidatedTool can forward them at runtime.
 // When multiple subcommands have the same flag name, the flag definition (type, description, etc.)
 // is taken from whichever subcommand is processed last, while AppliesToSubcommands tracks all
 // subcommands that use this flag. For consistent behavior, flags with the same name should have
 // the same type and description across subcommands.
-func mergeSubcommandFlags(subcommands map[string]*SubcommandDef) map[string]*FlagDef {
+func mergeSubcommandFlags(
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go:327]: Fixed in cb4d28cdd. `mergeSubcommandFlags` now sorts subcommand keys before iterating so the result is deterministic — the first alphabetical key wins when the same flag appears in multiple subcommands. The doc comment has been updated to reflect this first-wins behavior.

File: pkg/toolgen/definition.go [hunk 25]
```diff
@@ -6,6 +6,10 @@ type ToolDefinition struct {
 	// Name is the tool identifier (e.g., "ksail_cluster_create").
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/definition.go]: Fixed in cb4d28cdd. The Name docstring now uses `"cluster_create"` and `"workload_read"` as examples, matching the actual generated names (root command prefix is stripped).

File: pkg/toolgen/consolidate_helpers_test.go [hunk 26]
```diff
@@ -254,3 +254,64 @@ func TestCollectAllSubcommands_DeepNesting(t *testing.T) {
 	require.Len(t, subcommands, 1)
 	assert.Contains(t, subcommands, "level1_level2_leaf")
 }
+
+// TestGenerateTools_ConsolidateWithParentPermission exercises the parentHasPermission
+// branch of commandToPermissionSplitTools.
+// When a command carries both AnnotationConsolidate and AnnotationPermission,
+// all subcommands collapse into a SINGLE tool instead of a read/write pair.
+// Adding an excluded flag to a subcommand also exercises the ExcludeFlags path
+// inside extractFlags during consolidation.
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/consolidate_helpers_test.go]: Fixed in cb4d28cdd. The function doc comment now says "exercises the schema-level exclusion path in mergeSubcommandFlags during consolidation" to match the actual code path after the ExcludeFlags filter was moved out of extractFlags.

File: pkg/toolgen/mcp.go [hunk 27]
```diff
@@ -73,6 +78,20 @@ func addMCPTool(server *mcp.Server, tool ToolDefinition, opts ToolOptions) {
 	mcp.AddTool(server, mcpTool, handler)
 }
 
+// buildMCPAnnotations creates MCP ToolAnnotations from the tool's behavioral hints.
+// DestructiveHint and OpenWorldHint use *bool in the MCP SDK to distinguish
+// "explicitly false" from "unset" (nil defaults to true per the spec).
+// ReadOnlyHint and IdempotentHint are plain bool (nil defaults to false = zero value).
+func buildMCPAnnotations(tool ToolDefinition) *mcp.ToolAnnotations {
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/mcp.go:86]: Fixed in cb4d28cdd. The comment now emphasizes that KSail always provides explicit values (nil is never used here), and clarifies that the `*bool` type comes from the MCP SDK's nil/false distinction per the spec.

File: pkg/toolgen/schema.go [hunk 28]
```diff
@@ -429,3 +483,38 @@ func mapFlagTypeToItemsType(flagType string) string {
...
+	}
+
+	// Reserve space for the ellipsis
+	cutoff := maxLen - len(ellipsis)
+
+	// Try to break at natural boundaries in descending preference
+	for _, sep := range []string{". ", "; ", ", "} {
+		if idx := strings.LastIndex(desc[:cutoff], sep); idx > 0 {
+			return desc[:idx+len(sep)-1] + ellipsis
+		}
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/schema.go:510]: Fixed in 9964a779c. The separator is now dropped entirely at the break point (`desc[:idx]` instead of `desc[:idx+len(sep)-1]`) so the output is `'First sentence...'` instead of `'First sentence....'`. Tests updated to match.

File: pkg/toolgen/mcp.go [hunk 29]
```diff
@@ -73,6 +78,21 @@ func addMCPTool(server *mcp.Server, tool ToolDefinition, opts ToolOptions) {
 	mcp.AddTool(server, mcpTool, handler)
 }
 
+// buildMCPAnnotations creates MCP ToolAnnotations from the tool's behavioral hints.
+// DestructiveHint and OpenWorldHint use *bool in the MCP SDK because nil and false
+// have different meanings per the MCP spec (nil means "unset/unknown"; false is explicit).
+// KSail always provides explicit values for all hints, so nil is never used here.
+// ReadOnlyHint and IdempotentHint are plain bool (nil defaults to false = zero value).
```
[the remark this answers is not in the document]
Comment (devantler) [reply pkg/toolgen/mcp.go]: Fixed in 9964a779c. Comment updated to 'plain bool whose zero value is false' — bools cannot be nil so the '(nil defaults to false = zero value)' phrasing was removed.
